### PR TITLE
Alerting: Change getEvaluatorForAlertRule to checkDatasourcePermissionsForRule

### PR DESCRIPTION
**Pull request by yuri-tceretian**
**What this PR does / why we need it**:
This PR updates method getEvaluatorForAlertRule to accept permissions evaluator and exit on the first negative result rather than generating a slice of string. This is more effective and will let us use it in other places.

### CODE REVIEW

**Review by alexweav — APPROVED** [review]
LGTM

**Review by joeblubaugh — APPROVED** [review]
LGTM - could you write a bit in the commit message what it is about this change that makes the code more reusable? That will be helpful for changelog spelunkers in the future.

File: pkg/services/ngalert/api/authorization.go [hunk 1]
```diff
@@ -172,16 +172,17 @@ func (api *API) authorize(method, path string) web.Handler {
...
 		if query.QueryType == expr.DatasourceType || query.DatasourceUID == expr.OldDatasourceUID {
 			continue
 		}
-		scopes = append(scopes, ac.EvalPermission(datasources.ActionQuery, dashboards.ScopeFoldersProvider.GetResourceScopeUID(query.DatasourceUID)))
+		if !evaluator(ac.EvalPermission(datasources.ActionQuery, dashboards.ScopeFoldersProvider.GetResourceScopeUID(query.DatasourceUID))) {
+			return false
+		}
 	}
-	return ac.EvalAll(scopes...)
+	return true
```
Comment (gotjosh) [inline pkg/services/ngalert/api/authorization.go:185]: You can simplify this whole call by:

```golang
return evaluator(ac.EvalPermission(datasources.ActionQuery, dashboards.ScopeFoldersProvider.GetResourceScopeUID(query.DatasourceUID)
```

File: pkg/services/ngalert/api/authorization.go [hunk 2]
```diff
@@ -172,16 +172,17 @@ func (api *API) authorize(method, path string) web.Handler {
 	panic(fmt.Sprintf("no authorization handler for method [%s] of endpoint [%s]", method, path))
 }
 
-// GetDatasourceScopesFromAlertRule extracts data source scopes from an alert rule
-func getEvaluatorForAlertRule(rule *ngmodels.AlertRule) ac.Evaluator {
-	scopes := make([]ac.Evaluator, 0, len(rule.Data))
+// checkDatasourcePermissionsForRule checks that user has access to all data sources
+func checkDatasourcePermissionsForRule(rule *ngmodels.AlertRule, evaluator func(evaluator ac.Evaluator) bool) bool {
```
Comment (gotjosh) [inline pkg/services/ngalert/api/authorization.go:176]: nit: I'm not a fan of the word "check" as a verb for authorization. An alternative could be `verify` or `authorize` as checking seems a bit of an ambiguous word.

**Review by gotjosh — APPROVED** [review]
LGTM, my comments are nits and I don't need to see this again.

**Pull request merged**